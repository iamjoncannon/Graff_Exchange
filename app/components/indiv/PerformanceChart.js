import React from "react";
import { connect } from "react-redux";
import {LineChart, Line, XAxis, YAxis, CartesianGrid, Tooltip, Legend} from 'recharts';
import {isDesktop} from '../utils'

class PerformanceChart extends React.Component {

    constructor(props){
        super(props)

        this.state = {}
    }

    componentDidMount() {
        
        this.resize()

        window.addEventListener("resize", this.resize.bind(this))

    }

    componentWillUnmount(){

        window.removeEventListener("resize", this.resize.bind(this))

    }

    componentDidUpdate(){

    }
    
    resize = () => {

        this.setState({
            dimensions: {
              width: this.container.offsetWidth,
              height: this.container.offsetHeight,
            },
          });
    }

    renderContent(){

        const { width, height } = this.state.dimensions;

        const  { selectedPortfolioItem, period } = this.props

        let data = selectedPortfolioItem.historical.slice(0, period)

        for(let each in data){
            data[each].date = data[each].date.replace("2019-", "").replace("2018-", "")
        }

        return (

            <LineChart 
                width={width} height={height} data={data}
                margin={{top: 30, right: 30, left: 30, bottom: 5}}
            >

                <XAxis dataKey="date"  hide={ !isDesktop() || period > 100 ? true : false}/>

                <YAxis domain={['auto', 'auto']} />
                
                {/* <CartesianGrid strokeDasharray="3 3"/> */}
                <Tooltip/>
                {/* <Legend /> */}
                <Line type="monotone" dataKey="open" stroke="blue" dot={false} /> 
                {/* activeDot={{r: 1}} */}
                <Line type="monotone" dataKey="high" stroke="green"dot={false} />
                <Line type="monotone" dataKey="low" stroke="red" dot={false} />
                <Line type="monotone" dataKey="close" stroke="black" dot={false}/>
          </LineChart>

        );
    } 

    render(){
      
        const { dimensions } = this.state;

        return(

            <div className="chart" ref={el => (this.container = el)}>

                { dimensions && this.props.selectedPortfolioItem.historical && this.renderContent()}

            </div>
        )
    }
}

const mapStateToProps = ({ Portfolio_state }) => {

    return {
      selectedPortfolioItem : Portfolio_state.selectedPortfolioItem
    };
};
  
export default connect(
    mapStateToProps,
    null
)(PerformanceChart);
