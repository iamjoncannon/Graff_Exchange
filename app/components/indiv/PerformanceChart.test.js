import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import PerformanceChart from './PerformanceChart'
import configureStore from 'redux-mock-store';
import thunk from "redux-thunk"
const mockStore = configureStore([thunk]);
import { BrowserRouter as Router } from "react-router-dom";

const fake_api_data = {"data":
    {"hydrate_time_series_data":
        [{
            "date":"2019-10-02",
            "open":1206.42,
            "high":1206.42,
            "low":1173.37,
            "close":1176.985,
            "__typename":"Time_Series_Data_Point"
        }]
    }
}

const fake_store_without_data =  {
 
    Portfolio_state: {
        portfolio: {
            FB: {
                symbol: "FB",
                price: 100,
                current_holding: 3
            }
        },
        selectedPortfolioItem: "FB"
    }
}

const fake_store_with_data =  {
 
    Portfolio_state: {
        portfolio: {
            FB: {
                symbol: "FB",
                price: 100,
                current_holding: 5,
                historical: fake_api_data.data.hydrate_time_series_data,
                data: {
                    changePercent: .03,
                    latestPrice: 200
                }
            }
        },
        selectedPortfolioItem: "FB"
    }
}

describe("PerformanceChart", ()=>{

    let component 
    let store 
    let connected_react_component

    beforeEach(()=>{
                
        store = mockStore( fake_store_with_data );
    
        component = renderer.create(
            <Provider store={store} >
               <Router >
                   <PerformanceChart pathname={"/"} />
               </Router>
           </Provider>)

        connected_react_component = component.root.children[0].children[0].children[0].children[0]._fiber.stateNode

    })

    it('1) renders successfully with data from redux store', ()=>{

        let tree = component.toJSON();
        
        expect(tree).toMatchSnapshot();
    })

    it("2) renders a LineChart componentwhen props.portfolio.historical is defined", ()=>{

        let chart = component.root.find(node => node.type.name === "CategoricalChartWrapper")

        expect(chart).toBeTruthy()
    })    
})


describe("PerformanceChart", ()=>{

    let component 
    let store 
    let connected_react_component

    beforeEach(()=>{
                
        store = mockStore( fake_store_without_data );
    
        component = renderer.create(
            <Provider store={store} >
               <Router >
                   <PerformanceChart pathname={"/"} />
               </Router>
           </Provider>)

        connected_react_component = component.root.children[0].children[0].children[0].children[0]._fiber.stateNode

    })
    
    it("3) does not render a LineChart when props.portfolio.historical is not defined", ()=>{

        let chart = component.root.findAll(node => node.type.name === "CategoricalChartWrapper")

        expect(chart.length).toEqual(0)
    })

    it("4) if props.portfolio.historical is not defined, renders Loading Dots", ()=>{

        let chart = component.root.find(node => node.type.name === "LoadingDots")

        expect(chart).toBeTruthy()

    })
    
})