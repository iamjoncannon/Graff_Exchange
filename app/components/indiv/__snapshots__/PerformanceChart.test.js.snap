// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PerformanceChart 1) renders successfully with data from redux store 1`] = `
<div
  className="chart"
>
  <div
    className="recharts-wrapper"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "cursor": "default",
        "height": 100,
        "position": "relative",
        "width": 100,
      }
    }
  >
    <svg
      className="recharts-surface"
      height={100}
      version="1.1"
      viewBox="0 0 100 100"
      width={100}
    >
      <defs>
        <clipPath
          id="recharts5-clip"
        >
          <rect
            height={65}
            width={-20}
            x={90}
            y={30}
          />
        </clipPath>
      </defs>
      <g
        className="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
      >
        <line
          className="recharts-cartesian-axis-line"
          fill="none"
          height={65}
          stroke="#666"
          width={60}
          x={30}
          x1={90}
          x2={90}
          y={30}
          y1={30}
          y2={95}
        />
        <g
          className="recharts-cartesian-axis-ticks"
        >
          <g
            className="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              className="recharts-cartesian-axis-tick-line"
              fill="none"
              height={65}
              stroke="#666"
              width={60}
              x={30}
              x1={84}
              x2={90}
              y={30}
              y1={95}
              y2={95}
            />
            <text
              className="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height={65}
              stroke="none"
              textAnchor="end"
              width={60}
              x={82}
              y={95}
            >
              <tspan
                dy="0.355em"
                x={82}
              >
                1170
              </tspan>
            </text>
          </g>
          <g
            className="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              className="recharts-cartesian-axis-tick-line"
              fill="none"
              height={65}
              stroke="#666"
              width={60}
              x={30}
              x1={84}
              x2={90}
              y={30}
              y1={78.75}
              y2={78.75}
            />
            <text
              className="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height={65}
              stroke="none"
              textAnchor="end"
              width={60}
              x={82}
              y={78.75}
            >
              <tspan
                dy="0.355em"
                x={82}
              >
                1180
              </tspan>
            </text>
          </g>
          <g
            className="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              className="recharts-cartesian-axis-tick-line"
              fill="none"
              height={65}
              stroke="#666"
              width={60}
              x={30}
              x1={84}
              x2={90}
              y={30}
              y1={62.5}
              y2={62.5}
            />
            <text
              className="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height={65}
              stroke="none"
              textAnchor="end"
              width={60}
              x={82}
              y={62.5}
            >
              <tspan
                dy="0.355em"
                x={82}
              >
                1190
              </tspan>
            </text>
          </g>
          <g
            className="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              className="recharts-cartesian-axis-tick-line"
              fill="none"
              height={65}
              stroke="#666"
              width={60}
              x={30}
              x1={84}
              x2={90}
              y={30}
              y1={46.25}
              y2={46.25}
            />
            <text
              className="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height={65}
              stroke="none"
              textAnchor="end"
              width={60}
              x={82}
              y={46.25}
            >
              <tspan
                dy="0.355em"
                x={82}
              >
                1200
              </tspan>
            </text>
          </g>
          <g
            className="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              className="recharts-cartesian-axis-tick-line"
              fill="none"
              height={65}
              stroke="#666"
              width={60}
              x={30}
              x1={84}
              x2={90}
              y={30}
              y1={30}
              y2={30}
            />
            <text
              className="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height={65}
              stroke="none"
              textAnchor="end"
              width={60}
              x={82}
              y={30}
            >
              <tspan
                dy="0.355em"
                x={82}
              >
                1210
              </tspan>
            </text>
          </g>
        </g>
      </g>
      <g
        className="recharts-layer recharts-line"
      >
        <g
          className="recharts-layer recharts-line-dots"
          clipPath={null}
        >
          <circle
            className="recharts-dot recharts-line-dot"
            cx={80}
            cy={35.81749999999988}
            fill="#fff"
            height={65}
            r={3}
            stroke="blue"
            strokeWidth={1}
            width={-20}
          />
        </g>
      </g>
      <g
        className="recharts-layer recharts-line"
      >
        <g
          className="recharts-layer recharts-line-dots"
          clipPath={null}
        >
          <circle
            className="recharts-dot recharts-line-dot"
            cx={80}
            cy={35.81749999999988}
            fill="#fff"
            height={65}
            r={3}
            stroke="green"
            strokeWidth={1}
            width={-20}
          />
        </g>
      </g>
      <g
        className="recharts-layer recharts-line"
      >
        <g
          className="recharts-layer recharts-line-dots"
          clipPath={null}
        >
          <circle
            className="recharts-dot recharts-line-dot"
            cx={80}
            cy={89.52375000000018}
            fill="#fff"
            height={65}
            r={3}
            stroke="red"
            strokeWidth={1}
            width={-20}
          />
        </g>
      </g>
      <g
        className="recharts-layer recharts-line"
      >
        <g
          className="recharts-layer recharts-line-dots"
          clipPath={null}
        >
          <circle
            className="recharts-dot recharts-line-dot"
            cx={80}
            cy={83.64937500000016}
            fill="#fff"
            height={65}
            r={3}
            stroke="black"
            strokeWidth={1}
            width={-20}
          />
        </g>
      </g>
    </svg>
    <div
      className="recharts-tooltip-wrapper"
      style={
        Object {
          "MozTransform": "translate(undefinedpx, undefinedpx)",
          "OTransform": "translate(undefinedpx, undefinedpx)",
          "WebkitTransform": "translate(undefinedpx, undefinedpx)",
          "msTransform": "translate(undefinedpx, undefinedpx)",
          "pointerEvents": "none",
          "position": "absolute",
          "top": 0,
          "transform": "translate(undefinedpx, undefinedpx)",
          "visibility": "hidden",
        }
      }
    >
      <div
        className="recharts-default-tooltip"
        style={
          Object {
            "backgroundColor": "#fff",
            "border": "1px solid #ccc",
            "margin": 0,
            "padding": 10,
            "whiteSpace": "nowrap",
          }
        }
      >
        <p
          className="recharts-tooltip-label"
          style={
            Object {
              "margin": 0,
            }
          }
        >
          
        </p>
      </div>
    </div>
  </div>
</div>
`;
